tests/integration/examples/utils.py

```diff
@@ -0,0 +1,250 @@
+    latter option is needed for Windows compatibility.
+    """
+    return (
+        [os.environ[SHELL_EXECUTABLE]] if SHELL_EXECUTABLE in os.environ else []
+    ) + [str(examples_dir / EXAMPLES_RUN_SCRIPT)]
+
+
+@contextmanager
+def run_example(
+    request: pytest.FixtureRequest,
+    name: str,
+    *args: str,
+    pipeline_name: Optional[str] = None,
+    run_count: Optional[int] = None,
+    step_count: Optional[int] = None,
+) -> Generator[Tuple[LocalExample, List[PipelineRunView]], None, None]:
+    """Runs the given examples and validates they ran correctly.
+
+    Args:
+        request: The pytest request object.
+        name: The name (=directory name) of the example.
+        *args: Additional arguments to pass to the example
+        pipeline_name: Validate that a pipeline with this name was registered.
+        run_count: Validate that this many pipeline runs were executed during
+            the example run.
+        step_count: Validate that the pipeline runs had this many steps.
+
+    Yields:
+        The example and the pipeline runs that were executed and validated.
+    """
+
+    # Root directory of all checked out examples
+    examples_directory = Path(__file__).parents[3] / "examples"
+
+    dst_dir = Path(os.getcwd())
+
+    # Copy all example files into the repository directory
+    copy_example_files(str(examples_directory / name), str(dst_dir))
+
+    now = datetime.utcnow()
+
+    # Run the example
+    example = LocalExample(name=name, path=dst_dir)
+    example.run_example_directly(*args)
+
+    runs: List[PipelineRunView] = []
+    if pipeline_name:
+        runs = wait_and_validate_pipeline_run(
+            pipeline_name=pipeline_name,
+            run_count=run_count,
+            step_count=step_count,
+            older_than=now,
+        )
+
+    yield example, runs
+
+    cleanup_docker = request.config.getoption("cleanup_docker", False)
+
+    if cleanup_docker:
+        # Clean up more expensive resources like docker containers, volumes and
+        # images, if any were created.
+
+        try:
+            from docker.client import DockerClient
+            from docker.errors import ImageNotFound
+
+            # Try to ping Docker, to see if it's installed and running
+            docker_client = DockerClient.from_env()
+            docker_client.ping()
+        except Exception:
+            # Docker is not installed or running
+            pass
+        else:
+            docker_client.containers.prune()
+            docker_client.volumes.prune()
+            try:
+                pipeline_image = docker_client.images.get(
+                    f"zenml:{pipeline_name}"
+                )
+                logging.debug(f"Removing Docker image {pipeline_name}")
+                docker_client.images.remove(pipeline_image.id)
+            except ImageNotFound:
+                pass
+
+            docker_client.images.prune()
+
+
+def wait_and_validate_pipeline_run(
```

Comment:
the logic of this confuses me quite a bit. So if I get this right, it works like this:
- before the function is called, N runs have already been started or are about to start
- now we iterate N times, and in each iteration M, we first wait for M runs to exist, then wait for the last M runs to be finished
- at the end, we check whether we indeed found N runs and whether all of those completed

This has several issues in my opinion:
- in each iteration, we check the same runs again and again
- when we check whether M runs have finished, those might not necessarily be the same ones we checked as started
- the checks at the very end seem redundant

Unless there is an important reason not to, I would suggest to simplify this into a single while loop that waits for N runs to exist or until `start_timeout` is over, and then another while loop that repeatedly checks the status of all those runs until all of them have status `COMPLETED` or until the `finish_timeout` is reached or until any run had status `FAILED`

Reply:
I put a lot of thought into this and the main reason for the complexity is this: not all orchestrators are synchronized and not all pipeline runs are one-shot. Some runs may be scheduled, others may be executed in sequence, so this function is equipped to handle all those cases). Each run may take a while to complete too, so waiting for all of them to start in a single shot of `start_timeout` timeout is not feasible.

The logic required to wait for N consecutive runs to complete or for the first to fail while at the same time cutting the wait time as short as possible is not as easy as you describe it, but it can probably be implemented in a less complicated manner. I'll settle for this implementation for now, if that's ok with you, simply because this is already battle-tested and re-implementing it will set me back too much. But there's a silver lining: we can learn from this and implement something better in the ZenML post-execution workflow code, something that we can use to replace this. You're more than welcome to give it a try there :)

Reply:
alright agreed, let's keep it for now so we can get this merged and potentially redesign it for post execution later 👍